File: Lintcode/165-Merge-Two-Sorted-List/MergeTwoSortedLists.java

//  Definition for ListNode.
class ListNode {
	int val;
	ListNode next;

	ListNode(int val) {
		this.val = val;
		this.next = null;
	}
}

public class MergeTwoSortedLists {
	/**
     * @param ListNode l1 is the head of the linked list
     * @param ListNode l2 is the head of the linked list
     * @return: ListNode head of linked list
     */
    public ListNode mergeTwoLists(ListNode l1, ListNode l2) {
        if (l1 == null) {
            return l2;
        }
        if (l2 == null) {
            return l1;
        }
        ListNode cur1 = l1;
        ListNode cur2 = l2;
        ListNode dummy = new ListNode(0);
        ListNode curRes = dummy;
        while (cur1 != null || cur2 != null) {
            if (cur1 == null || (cur2 != null && cur1.val > cur2.val)) {
                curRes.next = cur2;
                cur2 = cur2.next;
            } else {
                curRes.next = cur1;
                cur1 = cur1.next;
            }
            curRes = curRes.next;
        }
        return dummy.next;
    }
}

File: Leetcode/21-Merge-Two-Sorted-Lists/MergeTwoSortedLists.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def mergeTwoLists(self, l1, l2):
        """
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """
        dummy = ListNode(0)
        cur = dummy
        while l1 != None and l2 != None:
            if l1.val < l2.val :
                cur.next = l1
                l1 = l1.next
            else :
                cur.next = l2
                l2 = l2.next
            cur = cur.next
        
        if l1 == None :
            cur.next = l2
        else :
            cur.next = l1
        return dummy.next;
        
